### DISCUSSION TITLE
Using a hook inside SWRConfiguration

### DISCUSSION THREAD

**Discussion by Jerryhu1** [vercel/swr#1644] [Q&A] [answered]
Hi,

I'm trying to combine a Zustand store with SWR. I wish to dispatch a Zustand action through `onLoadingSlow`, `onSuccess` and `onError`, but this gives me the following error: `Invalid hook call. Hooks can only be called inside of the body of a function component.` Is this possible somehow? 

Minimal repro codesandbox: https://codesandbox.io/s/lucid-pasteur-fin3w

I used to do this with a Redux dispatch which worked fine, but now when refactoring to Zustand it does not seem to be possible anymore. Here's the code: 

```
export const SWRGlobalConfig: SWRConfiguration = {
  errorRetryCount: 3,
  loadingTimeout: 2000,
  onLoadingSlow: () => {
    useDisplayState(state => state.activateLoader)();
  },
  onSuccess: () => {
    useDisplayState(state => state.deactivateLoader)();
  },
  onError: err => {
    const e = err as AxiosError;
    if (e.isAxiosError) {
      const addNotification = useNotificationStore(state => state.addNotification);
      addNotification({
        key: uuidv4(),
        message: e.message,
        severity: "error",
      });
    } else {
      const addNotification = useNotificationStore(state => state.addNotification);
      addNotification({
        key: uuidv4(),
        message: err,
        severity: "error",
      });
    }
    useDisplayState(state => state.deactivateLoader)();
  },
};
```

Help is greatly appreciated!

**Comment by huozhi** [comment, marked as answer]
Does defining the dispatcher above SWRConfiguration work for you? like

```js
const showLoading = useDisplayState(state => state.activateLoader)
export const SWRGlobalConfig = {
  onLoadingSlow: () => { showLoading(); },
  // ...
```

You cannot use hooks inside those callbacks since they're just callbaks triggered on state change, which is outside of react render scope.

**Reply by Jerryhu1** [reply]
Sadly, this gives me the same error, since i'm now calling a hook outside a function. Any other alternatives/ideas?

**Reply by huozhi** [reply]
could you share a codesandbox with a minimal repro that we can check what's going on? 🙏

**Reply by Jerryhu1** [reply]
First time doing this, here you go!: https://codesandbox.io/s/lucid-pasteur-fin3w

**Reply by nielskrijger** [reply]
Being unfamiliar with useSWR and Zustand, my first thought was to write a custom hook that wraps `useSWR` and do any global state updates there instead.

Something like this: https://codesandbox.io/s/green-wildflower-r5kcc?file=/src/App.tsx

**Reply by huozhi** [reply]
Yeah @nielskrijger is right, the problem in the reproduction given above is: you need to wrap hooks into functional component, and then passdown the handlers return from hooks to swr configuration.

```js
import useDisplayState from "./store/display";

function ConfiguredApp() {
  const show = useDisplayState((state) => state.activateLoader);
  const hide = useDisplayState((state) => state.deactivateLoader);
  const config: SWRConfiguration = {
    errorRetryCount: 3,
    loadingTimeout: 2000,
    onLoadingSlow: show,
    onSuccess: hide,
    onError: hide
  };
  return (
    <SWRConfig value={config}>
      <App />
    </SWRConfig>
  );
}

const rootElement = document.getElementById("root");

render(<ConfiguredApp />, rootElement);
```

You cannot directly call it outside of react render tree

**Reply by Jerryhu1** [reply]
Makes a lot of sense. Both solutions work, thanks!

**Comment by Jerryhu1** [comment]
As an additional answer to this, I found out `zustand` has a non-hook method of accessing the store as well by using `useStore().getState()`, the following code works without adding it to the FC: 

```
const { addNotification } = useNotificationState.getState()
const { activateLoader, deactivateLoader } = useDisplayState.getState()

export const SWRGlobalConfig: SWRConfiguration = {
  errorRetryCount: 3,
  loadingTimeout: 2000,
  onLoadingSlow: () => {
    activateLoader()
  },
  onSuccess: () => {
    deactivateLoader()
  },
  onError: err => {
    // TODO: add metrics for client side failures
    // by reporting to some backservice
    const e = err as AxiosError;
    if (e.isAxiosError) {
      addNotification({
        key: uuidv4(),
        message: e.message,
        severity: "error",
      })
    } else {
      addNotification({
        key: uuidv4(),
        message: err,
        severity: "error",
      });
    }
    deactivateLoader();
  },
};
```
